[caffe2] Fix compilation with fmt 8.x

vitaut opened the pull request:
Summary:
Fix a few issues that block migration to fmt 8.x:

1. Format strings must be known at compile time by default
2. `formatter` specialization must be visible when formatting an object

Test Plan: sandcastleit

Differential Revision: D33835157

suo commented:
Try rebasing on latest master to make the build errors go away.

vitaut commented:
@suo, rebased and the errors did go away, thanks.

cbalioglu commented on torch/csrc/distributed/c10d/logging.h:
```diff
@@ -10,11 +10,18 @@
 
 #include <c10/util/Logging.h>
 
+namespace torch {
```
Nit: do you mind changing the namespace to `c10d::detail` and naming the function `vformat`?

```
namespace c10d {
namespace detail {

template <typename... T>
std::string vformat(fmt::string_view fmt, T&&... args) {
  return fmt::vformat(fmt, fmt::make_format_args(std::forward<Args>(args)...));
}

} // namespace detail
} // namespace c10d
```

vitaut replied:
Done.

cbalioglu commented on torch/csrc/distributed/c10d/logging.h:
```diff
@@ -10,11 +10,18 @@
+namespace torch {
+template <typename... T>
+std::string c10d_format(fmt::string_view fmt, T&&... args) {
+  return fmt::vformat(fmt, fmt::make_format_args(args...));
```
I believe you should use `std::forward<Args>(args)...` instead of plain `args` to preserve the value categories of the argument expressions.

vitaut replied:
Forwarding is not needed in `make_format_args`, only in formatting functions.

vitaut replied:
Actually you are right, `make_format_args` could use forwarding here in newer versions of fmt and it won't harm in older, so changed.

cbalioglu reviewed (commented):
Just to minor comments.

vitaut committed:
[caffe2] Fix compilation with fmt 8.x (#71966)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71966

Fix a few issues that block migration to fmt 8.x:

1. Format strings must be known at compile time by default
2. `formatter` specialization must be visible when formatting an object

Test Plan: sandcastleit

Differential Revision: D33835157

fbshipit-source-id: 88dfa1ddf30765f9f7d65c9a322a20d21bdefb15

cbalioglu reviewed (approved):
LGTM!
